core/drivers/amd/asu/asu_main.c

```diff
@@ -0,0 +1,522 @@
+{
+	if (!param_ptr) {
+		EMSG("Invalid Client Parameters");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (!param_ptr->cbptr) {
+		EMSG("Callback function not available");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
+	if (param_ptr->priority != ASU_PRIORITY_HIGH &&
+	    param_ptr->priority != ASU_PRIORITY_LOW) {
+		EMSG("Invalid Priority");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * send_doorbell() - Send IPI doorbell interrupt to ASU
+ *
+ * Triggers an Inter-Processor Interrupt (IPI) to notify the ASU
+ * that a new command is available in the shared memory queue.
+ *
+ * Return: TEE_SUCCESS on successful doorbell trigger
+ */
+static TEE_Result send_doorbell(void)
+{
+	io_write32((vaddr_t)asu->doorbell + IPIPSU_TRIG_OFFSET,
+		   ASU_TARGET_IPI_INT_MASK);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * asu_update_queue_buffer_n_send_ipi() - Queue command and send IPI
+ * @param: Client parameters including priority
+ * @req_buffer: Request buffer containing command data
+ * @size: Size of request data
+ * @header: Command header information
+ * @status: FW return status
+ *
+ * Places a command in the appropriate priority queue buffer, updates
+ * queue status, and sends an IPI to notify ASU of the pending command.
+ *
+ * Return: TEE_SUCCESS on success, appropriate error code on failure
+ */
+TEE_Result asu_update_queue_buffer_n_send_ipi(struct asu_client_params *param,
+					      void *req_buffer,
+					      uint32_t size,
+					      uint32_t header,
+					      int *status)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t freeindex = 0;
+	struct asu_channel_queue_buf *bufptr = NULL;
+	struct asu_channel_queue *qptr = NULL;
+
+	if (!param || header == 0U) {
+		EMSG("Invalid parameters provided");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (asu->is_ready != ASU_CLIENT_READY) {
+		EMSG("ASU Client is not ready");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	freeindex = get_free_index(param->priority);
+	if (freeindex == ASU_MAX_BUFFERS) {
+		EMSG("ASU Buffers Full");
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	if (param->priority == ASU_PRIORITY_HIGH) {
+		bufptr = &asu->chnl_memptr->p0_chnl_q.queue_bufs[freeindex];
+		qptr = &asu->chnl_memptr->p0_chnl_q;
+	} else {
+		bufptr = &asu->chnl_memptr->p1_chnl_q.queue_bufs[freeindex];
+		qptr = &asu->chnl_memptr->p1_chnl_q;
+	}
+
+	bufptr->req.header = header;
+	if (req_buffer && size != 0U)
+		memcpy(bufptr->req.arg, req_buffer, size);
+
+	bufptr->respbufstatus = 0;
+
+	qptr->cmd_is_present = true;
+	qptr->req_sent++;
+	ret = send_doorbell();
+	if (ret != TEE_SUCCESS) {
+		EMSG("Failed to communicate to ASU");
+		return TEE_ERROR_COMMUNICATION;
+	}
+	while (io_read8((vaddr_t)&bufptr->respbufstatus) !=
+	       ASU_RESPONSE_IS_PRESENT) {
+		/*
+		 * WFE will return on SEV generated by the
+		 * interrupt handler or by a spin_unlock
+		 */
+		wfe();
+	}
+
+	*status = bufptr->resp.arg[ASU_RESPONSE_STATUS_INDEX];
```

Comment:
bufptr->resp.arg[] is an array of `uint32_t`, why not make `status` a `uint32_t` pointer?